spec/controllers/omniauth_callbacks_controller_spec.rb

```diff
@@ -0,0 +1,95 @@
+require 'rails_helper'
+
+RSpec.describe OmniauthCallbacksController, type: :controller do
+  describe '#github' do
+    context 'when github email doesn\'t exist' do
+      before do
+        github_omniauth
+        get :github
+        @user = User.find_by(email: 'user@email.com')
+      end
+
+      it 'expects omniauth.auth to be be_truthy' do
+        expect(request.env['omniauth.auth']).to be_truthy
+      end
+
+      it 'creates authentication with github' do
+        authentication = @user.identities.find_by(provider: 'github', uid: '12345')
+        expect(authentication).to_not be_nil
+      end
+
+      it 'redirects to home page' do
+        expect(response).to redirect_to root_path
+      end
+    end
+
+    context 'when github email already exist' do
+      before do
+        github_omniauth
+        get :github
+      end
+
+      it { expect(request.env['omniauth.auth']).to be_truthy }
+
+      it 'finds user whith this email' do
+        user = Identity.find_by(provider: 'github', uid: '12345').user
+        expect(user).to_not be_nil
+      end
+
+      it 'redirects to home page' do
+        expect(response).to redirect_to root_path
+      end
+    end
+  end
+
+  describe '#google_oauth2' do
+    context 'when google email doesn\'t exist' do
+      before do
+        google_omniauth
+        get :google_oauth2
+        @user = User.find_by(email: 'user@email.com')
+      end
+
+      it { expect(request.env['omniauth.auth']).to be_truthy }
+
+      it 'creates authentication with google' do
+        authentication = @user.identities.find_by(provider: 'google', uid: '12345')
+        expect(authentication).to_not be_nil
+      end
+
+      it 'redirects to home page' do
+        expect(response).to redirect_to root_path
+      end
+    end
+
+    context 'when google email already exist' do
+      before do
+        google_omniauth
+        get :google_oauth2
+      end
+
+      it { expect(request.env['omniauth.auth']).to be_truthy }
+
+      it 'finds user whith this email' do
+        user = Identity.find_by(provider: 'google', uid: '12345').user
+        expect(user).to_not be_nil
+      end
+
+      it 'redirects to home page' do
+        expect(response).to redirect_to root_path
+      end
+    end
+  end
+end
+
+def github_omniauth
```

Comment:
This is not a good thing to use pure methods in specs. Code directly in `before` blocks if needed.